cirq/work/pauli_string_collector.py

```diff
@@ -0,0 +1,99 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import random
+from typing import Optional
+
+import numpy as np
+
+from cirq import circuits, study, ops, value
+from cirq.work import collector
+
+
+class PauliStringSampleCollector(collector.SampleCollector[ops.PauliString,
+                                                           float]):
+    """Estimates the energy of a linear combination of Pauli observables."""
+
+    def __init__(self,
+                 circuit: circuits.Circuit,
+                 terms: value.LinearDict[ops.PauliString],
+                 samples_per_job: int = 1000):
+        """
+        Args:
+            circuit: Produces the state to be tested.
+            terms: The pauli product observables to measure. Their sampled
+                expectations will be scaled by their coefficients and their
+                dictionary weights, and then added up to produce the final
+                result.
+            samples_per_job: How many samples to request at a time.
+        """
+        self._circuit = circuit
+        self._samples_per_job = 1000
+        self._terms = value.LinearDict({
+            abs(k): v * k.coefficient for k, v in terms.items()
+        })
+        self._zeros = collections.defaultdict(lambda: 0)
+        self._ones = collections.defaultdict(lambda: 0)
+
+    def next_job(self) -> Optional[collector.CircuitSampleJob[ops.PauliString]]:
+        pauli = random.choice(list(self._terms.keys()))
+        return collector.CircuitSampleJob(
+            circuit=_circuit_plus_pauli_string_measurements(self._circuit,
+                                                            pauli),
+            repetitions=self._samples_per_job,
+            key=pauli)
+
+    def on_job_result(self, job: collector.CircuitSampleJob, result: study.TrialResult):
+        parities = result.histogram(
+            key='out',
+            fold_func=lambda bits: np.sum(bits) % 2)
+        self._zeros[job.key] += parities[0]
+        self._ones[job.key] += parities[1]
+
+    def result(self) -> float:
+        energy = 0
+        for term, coef in self._terms.items():
+            a = self._zeros[term]
+            b = self._ones[term]
+            if a + b:
+                energy += coef * (a - b) / (a + b)
+        return energy
+
+
+def _circuit_plus_pauli_string_measurements(circuit: circuits.Circuit,
+                                            pauli_string: ops.PauliString
+                                            ) -> circuits.Circuit:
+    """A circuit measuring the given observable at the end of the given circuit.
+    """
+    assert pauli_string
+    n = len(pauli_string.keys())
+    measure_layer = ops.Moment([
+        ops.measure(*sorted(pauli_string.keys()), key='out')
+    ])
+
+    circuit = circuit.copy()
+    fix_map = {
+        ops.X: ops.Y**0.5,
+        ops.Y: ops.X**-0.5,
+    }
+    transform_layer = ops.Moment([
```

Comment:
nit: seems like the fix_map code should live over in pauli land.  This is a common thing to want.

Reply:
It actually already exists as `PauliString.to_z_basis_ops`.

Reply:
That's much better.